vyper/codegen/module.py

```diff
@@ -160,15 +166,35 @@ def _runtime_ir(runtime_functions, all_sigs, global_ctx):
     # fallback label is the immediate next instruction,
     close_selector_section = ["goto", "fallback"]
 
-    runtime = [
-        "seq",
-        # check that calldatasize is at least 4, otherwise
-        # calldataload will load zeros (cf. yellow paper).
-        ["if", ["lt", "calldatasize", 4], ["goto", "fallback"]],
-        ["with", "_calldata_method_id", shr(224, ["calldataload", 0]), selector_section],
-        close_selector_section,
-        ["label", "fallback", ["var_list"], fallback_ir],
-    ]
+    runtime = ["seq"]
+
+    # if there is a function whose selector is 0, this gets triggered
+    # instead of the fallback fn when 0 calldataload is supplied,
+    # b/c calldataload loads 0s past the end of physical calldata (cf. yellow paper),
+    # causing the 0 selector fn to be treated as the fallback fn.
+    # since supplying 0 calldata is expected to trigger the fallback fn,
+    # we check that calldatasize >= 4, which distinguishes the 0 selector
+    # from the fallback function "selector"
+    # (equiv. to "all selectors not in the selector table").
+    # for all other fns, calldatasize is already validated at entry
+    # to the fn to be at least >= 4, so we can elide the check.
+
+    # note: this check is also present for fns with a 0 selector,
+    # but the check is an assertion, rather than jumping to fallback
+    # if the assertion fails.
+    # note: fns with trailing 0s in the selector are not checked here
+    # b/c not providing the trailing 0s in calldata is considered bad practice,
+    # and the contract should revert.
+    if has_zero_selector_functions:
+        runtime.append(["if", ["lt", "calldatasize", 4], ["goto", "fallback"]])
```

Comment:
That is a good idea. Here are the cases I've come up with:
When 0x is supplied:
  the contract will go to fallback
If 0x0 is supplied:
  This is not valid bytecode and should not get accepted on the client's side
When 0x00 is supplied, or anything less than 4 bytes:
   the contract will revert in the selector table check
    (I can add an assert that calldata_min_size >= 4 to double check that args_abi_t.min_size() is not negative)
When 0x00000000 is supplied:
   The check in selector table will pass and then the 0 selector function will get appropriately treated as the 0 selector fn

Can you think of any others? If not I think this is safe to implement

Reply:
actually, idk. it incurs a calldatasize check at the beginning of the selector function. maybe cleaner is to add the check in the 0 selector entry itself. this will also push all the invariants into the same place in the codegen, so we don't need to reason about the invariants in different places going forward.

Reply:
so a carveout in the selector table -- if the selector is 0, instead of generating the usual `if (eq _method_id <method id literal>)`, we have `if (and (eq _method_id <method id literal, aka 0>) (gt calldatasize 0))`

Reply:
I implemented this in the latest commit -- is this what you were going for?